src/sha256.c

```diff
@@ -0,0 +1,406 @@
+        CHANGED_BAG(result);
+    }
+    return result;
+}
+
+Obj FuncGAP_SHA256_HMAC(Obj self, Obj key, Obj text)
+{
+    UInt           i, klen;
+    UInt1          k_ipad[64], k_opad[64];
+    UInt1          digest[32];
+    sha256_state_t st;
+    Obj            result;
+
+    RequireStringRep(SELF_NAME, key);
+    RequireStringRep(SELF_NAME, text);
+
+    memset(k_ipad, 0x36, sizeof(k_ipad));
+    memset(k_opad, 0x5c, sizeof(k_opad));
+
+    klen = GET_LEN_STRING(key);
+    if (GET_LEN_STRING(key) > 64) {
+        sha256_init(&st);
+        sha256_update(&st, CHARS_STRING(key), klen);
+        sha256_final(&st);
+
+        be32encode(digest, st.r, sizeof(digest));
+        klen = 32;
+
+        for (i = 0; i < klen; i++) {
+            k_ipad[i] ^= digest[i];
+            k_opad[i] ^= digest[i];
+        }
+    }
+    else {
+        for (i = 0; i < klen; i++) {
+            k_ipad[i] ^= CHARS_STRING(key)[i];
+            k_opad[i] ^= CHARS_STRING(key)[i];
+        }
+    }
+
+    sha256_init(&st);
+    sha256_update(&st, k_ipad, 64);
+    sha256_update(&st, CHARS_STRING(text), GET_LEN_STRING(text));
+    sha256_final(&st);
+
+    be32encode(digest, st.r, sizeof(digest));
+    sha256_init(&st);
+    sha256_update(&st, k_opad, 64);
+    sha256_update(&st, digest, 32);
+    sha256_final(&st);
+
+    result = NEW_PLIST(T_PLIST, 8);
+    SET_LEN_PLIST(result, 8);
+    for (i = 0; i < 8; i++) {
+        SET_ELM_PLIST(result, i + 1, ObjInt_UInt(st.r[i]));
+        CHANGED_BAG(result);
+    }
+    return result;
+}
+
+// Table of functions to export
+static StructGVarFunc GVarFuncs[] = {
+    GVAR_FUNC(GAP_SHA256_INIT, 0, ""),
+    GVAR_FUNC(GAP_SHA256_UPDATE, 2, "state, bytes"),
+    GVAR_FUNC(GAP_SHA256_FINAL, 1, "state"),
+    GVAR_FUNC(GAP_SHA256_HMAC, 2, "key, text"),
+
+    { 0 }    // Finish with an empty entry
+};
+
+/****************************************************************************
+**
+*F  InitKernel( <module> ) . . . . . . .  initialise kernel data structures
+*/
+static Int InitKernel(StructInitInfo * module)
+{
+    ImportGVarFromLibrary("GAP_SHA256_State_Type", &GAP_SHA256_State_Type);
+
+    // init filters and functions
+    InitHdlrFuncsFromTable(GVarFuncs);
+
+    // return success
+    return 0;
+}
+
+/****************************************************************************
+**
+*F  InitLibrary( <module> ) . . . . . . .  initialise library data structures
+*/
+static Int InitLibrary(StructInitInfo * module)
+{
+    // init filters and functions
+    InitGVarFuncsFromTable(GVarFuncs);
+
+    return 0;
+}
+
+/****************************************************************************
+**
+*F  InitSHA256() . . . . . . . . . . . . . . . . . .  table of init functions
+*/
+static StructInitInfo module = {
+    // init struct using C99 designated initializers; for a full list of
+    // fields, please refer to the definition of StructInitInfo
+    .type = MODULE_DYNAMIC,
```

Comment:
this needs updating now it's internal code

Reply:
indeed